ts/nni_manager/training_service/reusable/gpuScheduler.ts

```diff
@@ -51,46 +53,134 @@ export class GpuScheduler {
 
     /**
      * Schedule a machine according to the constraints (requiredGPUNum)
-     * @param requiredGPUNum required GPU number
+     * @param defaultRequiredGPUNum the default required GPU number when constraint.type === 'None'
      */
-    public scheduleMachine(environments: EnvironmentInformation[], requiredGPUNum: number | undefined, trialDetail: TrialDetail): GpuScheduleResult {
-        if (requiredGPUNum === undefined) {
-            requiredGPUNum = 0;
-        }
-        assert(requiredGPUNum >= 0);
-        // Step 1: Check if required GPU number not exceeds the total GPU number in all machines
-        const eligibleEnvironments: EnvironmentInformation[] = environments.filter((environment: EnvironmentInformation) =>
-            environment.defaultGpuSummary === undefined || requiredGPUNum === 0 || (requiredGPUNum !== undefined && environment.defaultGpuSummary.gpuCount >= requiredGPUNum));
-        if (eligibleEnvironments.length === 0) {
-            // If the required gpu number exceeds the upper limit of all machine's GPU number
-            // Return REQUIRE_EXCEED_TOTAL directly
-            return ({
-                resultType: ScheduleResultType.REQUIRE_EXCEED_TOTAL,
+    public scheduleMachine(environments: EnvironmentInformation[], constraint: PlacementConstraint,
+        defaultRequiredGPUNum: number | undefined, trialDetail: TrialDetail): GpuScheduleResult {
+        if (constraint.type == 'None' || constraint.type == 'GPUNumber') {
+            let requiredGPUNum = 0;
+            if (constraint.type == 'None') {
+                if (defaultRequiredGPUNum === undefined) {
+                    requiredGPUNum = 0;
+                } else {
+                    requiredGPUNum = defaultRequiredGPUNum;
+                }
+            } else if (constraint.type == 'GPUNumber') {
+                const gpus = constraint.gpus as Array<number>;
+                if (gpus.length != 1) {
```

Comment:
If `constraint.type == 'GPUNumber'` only support 1 gpu number, seems it does not have difference with `constraint.type == None`, why not just set requiredGPUNum= this.config.trialGpuNumber? Can `constraint.type == 'GPUNumber` be removed?

Reply:
It allows different trials to use different number of GPUs by setting `constraint.type == 'GPUNumber'`.
`constraint.type == None` will force all trials to use the same number of GPUs, i.e., `this.config.trialGpuNumber`.

Supporting 1 GPU number is due to limitation of `GpuScheduleResult`, which only support one `EnvironmentInformation`  for now. I use a list here for extensibility when we support multiple multiple environments.